work in progress for review

mjkkirschner opened the pull request:
WIP - definitely cleanup required but this "works", as in we can pass data to the pixelshader this way and points update in response. It needs:

- [ ] 1. cleanup, 
- [ ] 2. refactor of duplicated code
- [ ] 3. figure out why the points are rectangles, not ellipses.
- [ ] 4. another shader for lines
- [ ] 5. find another unused property for lines.

mjkkirschner committed:
work in progress for review
update shaders
use the blendFactor prop on pointGeometry to pack our data
material classes are not used

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Watch3D/shaderSource/vsDynamoPointLine.hlsl:
```diff
@@ -5,17 +5,17 @@
+GSInputPS main(VSInputPS input)
 {
-    PSInputPS output;
+	GSInputPS output;
```
Are you using the default geometry shader?

aparajit-pratap replied:
Did you have to make any changes to `DynamoEffectsManager`?

mjkkirschner replied:
I already added the geometry shader in the last commits that you asked me to commit directly to your branch.

mjkkirschner replied:
also, I used the point geometry shader that helix uses by default - for lines we'll need to add another shader and use the line geometry shader from helix.

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Watch3D/DynamoPointLineMaterialVariable.cs:
```diff
@@ -165,8 +166,115 @@ public override MaterialVariable CreateMaterialVariables(IEffectsManager manager
+        /// <summary>
+        /// Is this model Frozen.
+        /// </summary>
+        public bool IsFrozenData { get { return isFrozenData; } internal set { SetAffectsRender(ref isFrozenData, value); } }
```
Is there something special about doing `SetAffectsRender`? I've seen `Set` used in some places and another `Set`something method in helix.

aparajit-pratap replied:
I see that you didn't have to override `OnUpdateModelStruct`; I remember you mentioned it doesn't get called anyway. Secondly, I'm still a little confused why you didn't need to add the geometry shader to the shader passes.

mjkkirschner replied:
`OnUpdateModelStruct` now gets called since the pipeline is setup correctly, unfortunately the point model struct is very small and doesn't have anywhere we can stick our extra data.

it's only 2 bools and a matrix that we need to do the projection - so I just used a value that is set to the PointLine constant buffer, I don't know how to update this otherwise, but it looks like it is somehow hooked into the default material inside the helix implementation ,but fortunately we can access it from the geometryModel object directly, so no need to hook into the render events etc.

I still really want to know how to add arbitrary data instead of reusing these unused variables, hopefully we can get more info from the helix maintainers for the future...

aparajit-pratap commented:
This is brilliant btw! 🚀

mjkkirschner commented:
I think all we need to do to get the points rendering as circles is to add:

```
    if (pfParams[2] == 1)
    {
        float len = length(input.t);
        if (len > 1.4142)
            discard;
    }
```

to your pixel shader. pfParams[2] or pfParams.z are the same thing and contain the figure type. `FigureType = 1` is a circle, this just gets the length of the texture coordinates which range from -1 to 1, but are actually different lengths in world space depending on the output of the GS. So this just checks the radius is under a certain relative size and discards pixels outside that radius,.

aparajit-pratap commented:
So we need to set `pfParams[2]` to 1 in `CommonBuffers.hlsl`? I thought the square shape of the points was coming from the default geometry shader and I tried replacing the `makequads` with a circle with little success: https://github.com/helix-toolkit/helix-toolkit/blob/66a4fa8bb56ec24a7c1b562225453f85987e08b1/Source/HelixToolkit.Native.ShaderBuilder/GS/gsPoint.hlsl#L55

On a side note I was curious whether replacing our current implementation of coloring points and lines with shaders (or blending) was even worth doing. I noticed that even with a million points, the current implementation sets their colors almost instantaneously with selection and isolation turned on and off very quickly. I didn't test this with lines though. Maybe the slowness that DYN-2702 refers to is seen with lines?

aparajit-pratap commented:
> I thought the square shape of the points was coming from the default geometry shader and I tried replacing the `makequads` with a circle with little success.

I think what's happening is that quads (squares) are being drawn by the geometry shader and then pixels at the corners of those squares are being discarded in the pixel shader (with the logic you shared) to give the effect of circular points.
![image](https://user-images.githubusercontent.com/5710686/83342824-c96d6900-a2c1-11ea-998f-0cf2a332919a.png)

mjkkirschner commented:
> > I thought the square shape of the points was coming from the default geometry shader and I tried replacing the `makequads` with a circle with little success.
> 
> I think what's happening is that quads (squares) are being drawn by the geometry shader and then pixels at the corners of those squares are being discarded in the pixel shader (with the logic you shared) to give the effect of circular points.
> ![image](https://user-images.githubusercontent.com/5710686/83342824-c96d6900-a2c1-11ea-998f-0cf2a332919a.png)


yep, that exactly what happens, the pixel shader clips the output from the geometry shader.

mjkkirschner commented:
`On a side note I was curious whether replacing our current implementation of coloring points and lines with shaders (or blending) was even worth doing. I noticed that even with a million points, the current implementation sets their colors almost instantaneously with selection and isolation turned on and off very quickly. I didn't test this with lines though. Maybe the slowness that DYN-2702 refers to is seen with lines?`


On my machine I notice this mostly when drag selecting over the nodes. It's painfully slow, maybe this is a parallels thing? Or just a slow machine thing? (my machine is a few years old)

lines will be at least twice as slow as points, as each line contains more than 1 point and each needs to have its vertex color updated separately.

I think the worst case scenario to test is to try this with points and lines that are using vertex colors ie - those drawn using Display.ByGeometryColor node.

mjkkirschner commented:
@aparajit-pratap giving a shot at a refactor here, lets try to catch up tomorrow.

aparajit-pratap commented:
@mjkkirschner will these changes affect the performance of `ByGeometryColor` as-is? As I see it in that node, we are still individually assigning the new color to each vertex, which is why it still seems slow. Is the goal to set one of the variables used in the pixel shader to the color selected in that node in order to see the benefits of the shader? I think we should be doing the same for meshes as well, don't we?

mjkkirschner commented:
I think a few subtleties are being confused.
The initial setting of each vertex color at mesh, or point, or line creation time is a small overhead compared to the runtime change that occurs when we say ask the HelixViewModel to reset the colors to their original colors before doing a freeze for example. Only points and lines suffer from this performance penalty because we're using the vertex colors directly.

We could make even more optimizations as you say, to all shaders to use the first vertex color to display as ALL vertex colors, this would let us only send a single color and have the GPU do the rest for us - but we would need to set it into a buffer somewhere, it would not make sense to have each vertex refer to the first vertex in the shader.

This would **not** work well for custom drawn meshes where users override `tessellate` and draw whatever colors they want... we still need to support that use case.
⚠️ 


Thats not something I'm looking into at the moment as I *think* it's an optimization not a major performance issue.

In my mind the biggest issue is the whole requestResetColor event and the fact that is exists at all.

mjkkirschner committed:
update shaders
find same buffer slot for lines and points yay
refactor
shared data core
update tests

mjkkirschner commented:
@aparajit-pratap - if you have a chance tomorrow I'd like to go over my changes here with you to get your feedback.

summary is as follow:
* I moved out the data stuff to a shared core, so that mesh, lines,and points all have an instance of this data since they all share the same properties (selection, freeze etc).
* I found that we can access the existing material slots for points and lines and so those can share the same core you added and the shaders can use the same slot - we still need different shaders though because the line shader will require a different GS shader (we should just use the existing helix one) - see how I used the existing point GS shader.
* I added the clipping if statement to the PSPoint shader.
* updated some tests to use the new data core.

I think all thats left here is implementing the line shader and hooking it up with the correct name.



![Screen Shot 2020-05-31 at 8 01 17 PM](https://user-images.githubusercontent.com/508936/83365995-42c89280-a37a-11ea-948e-39f99cfce8d3.png)

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/Watch3D/DynamoLineGeometryModel3D.cs:
```diff
@@ -13,6 +13,17 @@ protected override SceneNode OnCreateSceneNode()
+        /// <param name="state"></param>
+        internal void SetState(int state)
+        {  
+             //TODO should we also use this for points if it works for lines to make them consistent?
```
this TODO is out of date, already did this.

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/Watch3D/DynamoPointGeometryModel3D.cs:
```diff
@@ -24,7 +31,30 @@ protected override RenderCore OnCreateRenderCore()
+        internal void SetPropertyData(DependencyPropertyChangedEventArgs args, GeometryModel3D geo)
+        {
+            this.dataCore.SetPropertyData(args);
+            //re reuse blending factor because our shader does not use it and we need space for an int
```
comment is out of date, will remove it.

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Watch3D/DynamoLineGeometryModel3D.cs:
```diff
@@ -13,6 +13,17 @@ protected override SceneNode OnCreateSceneNode()
+        internal void SetState(int state)
+        {  
+             //TODO should we also use this for points if it works for lines to make them consistent?
+            this.material.FadingNearDistance = state;
```
Did you replace this for `blendingfactor` as it is common to both point and line and because `blendingfactor` is missing for line?

mjkkirschner replied:
yep yep.

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Watch3D/DynamoLineGeometryModel3D.cs:
```diff
@@ -13,6 +13,17 @@ protected override SceneNode OnCreateSceneNode()
         {
             return new DynamoLineNode();
```
Don't we have to set `{ Material = material; }` in the constructor of `DynamoLineNode` as well?

mjkkirschner replied:
yes, will fix that, thanks for catching it.

aparajit-pratap commented:
@mjkkirschner I think this looks good. Thanks for the hard work on it to bring it to some final shape.

It would be interesting to see if there are any perceivable performance improvements in selection/frozen etc. state changes on point as lines as the current performance still doesn't seem all that bad to me. The following is from the build we used for the helix bug bash. The graph has about 50,000 points and 810,000 lines:
![geom_select_many](https://user-images.githubusercontent.com/5710686/83366585-0f880280-a37e-11ea-9d61-bd6e695169de.gif)

aparajit-pratap commented:
Don't we also have to set pfParams[2] to 1 in CommonBuffers.hlsl for the `Figure` type to be chosen to be a circle: https://github.com/helix-toolkit/helix-toolkit/blob/76dc523c9bf928ef2c6f6e72246df28959cc0336/Source/HelixToolkit.Native.ShaderBuilder/Common/CommonBuffers.hlsl#L145?

mjkkirschner commented:
> Don't we also have to set pfParams[2] to 1 in CommonBuffers.hlsl for the figure to be chosen to be a circle: https://github.com/helix-toolkit/helix-toolkit/blob/76dc523c9bf928ef2c6f6e72246df28959cc0336/Source/HelixToolkit.Native.ShaderBuilder/Common/CommonBuffers.hlsl#L145?


no, pfParams is set by the default point material based on the height, width, figure, and figure ratio properties that the PointGeometryModel have been set to.
https://github.com/helix-toolkit/helix-toolkit/blob/76dc523c9bf928ef2c6f6e72246df28959cc0336/Source/HelixToolkit.SharpDX.Shared/Model/Material/Variables/PointMaterialVariable.cs#L55


and we already set the pointFigure to circle when we create DynamoPointGeometryModels

mjkkirschner commented:
see here:
https://github.com/DynamoDS/Dynamo/blob/37514ba43bb802743555faa8460c45fad807c63c/src/DynamoCoreWpf/ViewModels/Watch3D/HelixWatch3DViewModel.cs#L2073

Ellipse is `1` as this is an enum.
